tests/integration/v3_failover_test.go

```diff
@@ -0,0 +1,173 @@
+// Copyright 2022 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package integration
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"go.etcd.io/etcd/api/v3/v3rpc/rpctypes"
+	clientv3test "go.etcd.io/etcd/tests/v3/integration/clientv3"
+	"testing"
+	"time"
+
+	"go.etcd.io/etcd/client/v3"
+	integration2 "go.etcd.io/etcd/tests/v3/framework/integration"
+	"google.golang.org/grpc"
+)
+
+func TestFailover(t *testing.T) {
+	cases := []struct {
+		name     string
+		testFunc func(*testing.T, *tls.Config, *integration2.Cluster) (*clientv3.Client, error)
+	}{
+		{
+			name:     "create client before the first server down",
+			testFunc: createClientBeforeServerDown,
+		},
+		{
+			name:     "create client after the first server down",
+			testFunc: createClientAfterServerDown,
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Logf("Starting test [%s]", tc.name)
+			integration2.BeforeTest(t)
+
+			// Launch an etcd cluster with 3 members
+			t.Logf("Launching an etcd cluster with 3 members [%s]", tc.name)
+			clus := integration2.NewCluster(t, &integration2.ClusterConfig{Size: 3, ClientTLS: &integration2.TestTLSInfo})
```

Comment:
does this TLS config match what is expected in production w.r.t. validity of serving certificates and distinct hostnames (or is it using localhost / identical loopback IPs for all members)? I have memories of the etcd client not handling TLS validation on failover correctly (e.g. https://github.com/etcd-io/etcd/pull/10911) and wanted to make sure we're actually exercising the scenario

Reply:
It's a good point, thanks for pointing out this.

It's using localhost & 127.0.0.1 (see below). The e2e & integration tests are supposed to be automatically executed in pipeline, so all etcd members are always running on the same host?

Probably we need to think about how to run & test etcd in an environment which is close to production? cc @ptabor @serathius @spzala 

```
        X509v3 extensions:
            X509v3 Key Usage: critical
                Digital Signature, Key Encipherment
            X509v3 Extended Key Usage: 
                TLS Web Server Authentication, TLS Web Client Authentication
            X509v3 Basic Constraints: critical
                CA:FALSE
            X509v3 Subject Key Identifier: 
                F4:B1:45:B9:C4:18:A5:74:3B:5A:C9:21:10:60:4F:6F:5F:EF:12:64
            X509v3 Authority Key Identifier: 
                keyid:29:26:2B:F4:ED:D9:18:F5:60:74:CE:00:7A:C0:E6:FF:1C:C8:BC:A0

            X509v3 Subject Alternative Name: 
                DNS:localhost, IP Address:127.0.0.1
```